File: src/main/java/tooko/pm/handlers/ChatPanel.java
package tooko.pm.handlers;

import cn.hutool.core.collection.*;
import cn.hutool.core.util.*;
import tooko.main.*;
import tooko.pm.*;
import tooko.td.TdApi.*;
import tooko.td.client.*;
import tooko.td.core.*;

import java.math.*;
import java.util.*;

public class ChatPanel extends TdHandler {

    public PmBot bot;
    public PmData data;

    public int DATA_ID = Fn.DataId._5;

    @Override
    public void onLoad() {

        bot = (PmBot) client;

        data = bot.data;

        initFunction("enter", "exit", "chat");

        initData(DATA_ID);

        initPayload("0");

    }

    @Override
    public void onPayload(User user, long chatId, Message message, String payload, String[] params) {

        send(Fn.deleteMessage(message));

        long sessionId = NumberUtil.isLong(params[0]) ? NumberUtil.parseLong(params[0]) : -1;

        User targetUser = user((int) sessionId);

        if (targetUser == null) {

            Fn.sendText(chatId, Fn.plainText(bot.OL.CHAT_NO_RECORD));

            return;

        }

        PmData.Session session = data.getSession(sessionId);

        send(Fn.sendText(chatId, message.replyToMessageId, sessionActions(session), Fn.parseHtml(sessionStat(session))));

    }


    @Override
    public void onFunction(User user, long chatId, Message message, String function, String param, String[] params, String[] originParams) {

        if (user.id != bot.bot.owner) {

            send(Fn.deleteMessage(message));

            return;

        }

        if ("enter".equals(function)) {

            long targetChat;

            if (message.replyToMessageId != 0) {

                targetChat = findChat(message);

                if (targetChat == -1) {

                    send(Fn.sendText(chatId, Fn.plainText(bot.OL.INVALID_REPLY)));

                    return;

                }

            } else if (params.length != 0) {

                User targetUser = findUser(message, param, params);

                if (targetUser == null) {

                    send(Fn.sendText(chatId, Fn.plainText(bot.OL.INVALID_CHAT_ID)));

                    return;

                }

                targetChat = targetUser.id;

            } else {

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.REPLY_OR_ID)));

                return;

            }

            Chat chat;

            try {

                chat = execute(new GetChat(targetChat));

            } catch (TdException e) {

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.GET_CHAT_FAILED, e)));

                return;

            }

            if (!(chat.type instanceof ChatTypePrivate)) {

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.NO_PRIVATE_WARN)));

                return;

            }

            bot.chat = targetChat;

            send(Fn.sendText(chatId, Fn.parseHtml(bot.OL.CHAT_ENTERED, Fn.mention(chat.title, (int) chat.id))));

        } else if ("exit".equals(function)) {

            if (bot.chat != -1) {

                bot.chat = -1;

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.CHAT_EXITED)));

            } else {

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.NO_CHAT_ENTERED)));

            }

        } else if ("chat".equals(function)) {

            long targetChat;

            if (message.replyToMessageId != 0) {

                targetChat = findChat(message);

                if (targetChat == -1) {

                    send(Fn.sendText(chatId, Fn.plainText(bot.OL.INVALID_REPLY)));

                    return;

                }

            } else if (params.length != 0) {

                User targetUser = findUser(message, param, params);

                if (targetUser == null) {

                    send(Fn.sendText(chatId, Fn.plainText(bot.OL.INVALID_CHAT_ID)));

                    return;

                }

                targetChat = targetUser.id;

            } else {

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.REPLY_OR_ID)));

                return;

            }

            PmData.Session session = data.getSession(targetChat);

            User targetUser = user((int) targetChat);

            if (targetUser == null) {

                send(Fn.sendText(chatId, Fn.plainText(bot.OL.CHAT_NO_RECORD)));

                return;

            }

            send(Fn.deleteMessage(message));

            send(Fn.sendText(chatId, message.replyToMessageId, sessionActions(session), Fn.parseHtml(sessionStat(session))));

        }

    }

    String sessionStat(PmData.Session session) {

        String stat = bot.OL.CHAT_MANAGE_HELP;

        stat += "\n\n" + bot.OL.USER_NAME + " : " + Fn.mention(user((int) session.chatId));

        if (data.blocked.contains(session.chatId)) {

            stat += " " + bot.OL.CHAT_IS_BLOCKED;

        }

        stat += "\n" + bot.OL.USER_ID + " : " + Fn.code(session.chatId);

        stat += "\n" + bot.OL.CHAT_MSG_RECEIVED;
        stat += "\n" + bot.OL.CHAT_MSG_SENDED;

        return StrUtil.format(stat, session.received.size(), session.sended.size());

    }

    InlineArray sessionActions(PmData.Session session) {

        byte[] sessionId = BigInteger.valueOf(session.chatId).toByteArray();

        InlineArray buttons = new InlineArray();

        buttons.dataLine(bot.OL.CHAT_MSG_DEL_RECEIVED, DATA_ID, 0, sessionId);
        buttons.dataLine(bot.OL.CHAT_MSG_DEL_SENDED, DATA_ID, 1, sessionId);
        buttons.dataLine(bot.OL.CHAT_MSG_DEL_ALL, DATA_ID, 2, sessionId);

        buttons.dataLine(data.blocked.contains(session.chatId) ? bot.OL.CHAT_UNBLOCK : bot.OL.CHAT_BLOCK, DATA_ID, 3, sessionId);

        buttons.dataLine(bot.OL.FINISH_MANAGE, DATA_ID, 4);

        return buttons;

    }

    @Override
    public void onNewCallbackQuery(int userId, long chatId, long messageId, long queryId, int subId, byte[][] data) {

        if (subId == 4) {

            send(Fn.deleteMessages(chatId, messageId));

            return;

        }

        long targetChat = new BigInteger(data[0]).longValue();

        PmData.Session session = this.data.getSession(targetChat);

        if (subId < 3) {

            if (targetChat == bot.current) bot.current = -1;

            HashSet<Long> sended = new HashSet<>();
            HashSet<Long> received = new HashSet<>();
            HashSet<Long> reports = new HashSet<>();

            if (subId > 0) {

                if (CollectionUtil.isEmpty(session.sended)) {

                    if (subId == 1) {

                        send(Fn.answerAlert(queryId, bot.OL.CHAT_EMPTY));

                        return;

                    }

                } else {

                    sended.addAll(session.sended);

                    for (long send : session.sended) {

                        reports.add(NumberUtil.parseLong(this.data.sended.removeValue(send)));

                    }

                    session.sended.clear();

                    if (subId != 2) {

                        this.data.sessions.setById(session.chatId, session);

                    }

                }

            }

            if (subId != 1) {

                if (CollectionUtil.isEmpty(session.received)) {

                    if (subId == 0) {

                        send(Fn.answerAlert(queryId, bot.OL.CHAT_EMPTY));

                        return;

                    }

                } else {

                    received.addAll(session.received);

                    for (long receive : session.received) {

                        reports.add(NumberUtil.parseLong(this.data.received.removeValue(receive)));

                    }

                    session.received.clear();

                    if (subId != 2) {

                        this.data.sessions.setById(session.chatId, session);

                    }

                }

            }

            HashSet<Long> marged = new HashSet<>();

            marged.addAll(received);
            marged.addAll(sended);

            if (marged.isEmpty()) {

                send(Fn.answerAlert(queryId, bot.OL.CHAT_EMPTY));

                return;

            }

            send(Fn.deleteMessages(session.chatId, Fn.toArray(marged)));

            String status = "";

            if (!received.isEmpty()) {

                status += StrUtil.format(bot.OL.PM_DELETED_RECEIVED, received.size());

                if (!sended.isEmpty()) status += "\n";

            }

            if (!sended.isEmpty()) {

                status += StrUtil.format(bot.OL.PM_DELETED_SENDED, sended.size());

            }

            send(Fn.answerAlert(queryId, status));

            if (subId == 2) {

                if (session.extras != null) reports.addAll(session.extras);

                if (session.reports != null) reports.addAll(session.reports);

                I(Fn.deleteMessages(bot.bot.owner, Fn.toArray(reports)));

                this.data.sessions.deleteById(session.chatId);

            } else {

                send(Fn.deleteMessages(bot.bot.owner, Fn.toArray(reports)));

            }

            I(Fn.editText(chatId, messageId, sessionActions(session), Fn.parseHtml(sessionStat(session))));

        } else if (subId == 3) {

            boolean target;

            if (this.data.blocked.contains(targetChat)) {

                this.data.blocked.remove(targetChat);

                target = false;

            } else {

                this.data.blocked.add(targetChat);

                target = true;

            }

            I(Fn.editText(chatId, messageId, sessionActions(session), Fn.parseHtml(sessionStat(session))));

            send(Fn.answerText(queryId, target ? bot.OL.CHAT_BLOCKED : bot.OL.CHAT_UNBLOCKED));

        }

    }

    User findUser(Message message, String param, String[] params) {

        if (NumberUtil.isInteger(param)) {

            return user(NumberUtil.parseInt(param));

        }

        for (TextEntity entity : ((MessageText) message.content).text.entities) {

            if (entity.type instanceof TextEntityTypeMentionName) {

                int userId = ((TextEntityTypeMentionName) entity.type).userId;

                return user(userId);

            }

        }

        String userName = params[0];

        try {

            if (userName.startsWith("@")) userName = userName.substring(1);

            Chat chat = execute(new SearchPublicChat(userName));

            if (chat.type instanceof ChatTypePrivate) {

                return user(((ChatTypePrivate) chat.type).userId);

            }

        } catch (TdException ignored) {
        }

        return null;

    }

    long findChat(Message message) {

        long actionMessage = message.replyToMessageId;

        String actionMessageStr = actionMessage + "";

        if (data.received.containsKey(actionMessageStr)) {

            return data.getSessionByElement("received", data.received.get(actionMessageStr)).chatId;

        } else if (data.sended.containsKey(actionMessageStr)) {

            return data.getSessionByElement("sended", data.sended.get(actionMessageStr)).chatId;

        } else {

            PmData.Session session = data.getSessionByElement("reports", actionMessage);

            if (session != null) {

                return session.chatId;

            }

            session = data.getSessionByElement("extras", actionMessage);

            if (session != null) {

                return session.chatId;

            }

        }

        return -1;

    }

}

File: old/pm/handlers/ChatPanel.kt
package tookox.pm.handlers

import cn.hutool.core.util.NumberUtil
import cn.hutool.core.util.StrUtil
import tooko.main.Fn
import td.TdApi.*
import tookox.core.client.TdException
import tookox.core.*
import tookox.core.client.*
import tookox.core.utils.*
import tookox.pm.PmBot
import tookox.pm.PmData
import kotlin.properties.Delegates

class ChatPanel : TdBotHandler() {

    override val sudo: PmBot get() = super.sudo as PmBot

    val bot get() = sudo.bot
    val data get() = sudo.data
    
    override fun onLoad() {

        initFunction("enter", "exit", "chat")

        initData(DATA_1)

        initStartPayload("0")
        
    }

    override suspend fun onFunction(userId: Int, chatId: Long, message: Message, function: String, param: String, params: Array<String>, originParams: Array<String>) {

        if (userId != bot.owner) rejectFunction()

        val L = userId.langForUserId
        
        if ("enter" == function) {
            
            var targetChat by Delegates.notNull<Long>()
            
            if (message.replyToMessageId != 0L) {
            
                targetChat = findChat(message)
            
                if (targetChat == -1L) {

                    sudo make L.INVALID_REPLY sendTo chatId
                    
                    return

                }
            
            } else if (params.isNotEmpty()) {
                
                val targetUser = findUser(message, param, params)
                
                if (targetUser == null) {

                    sudo make L.INVALID_CHAT_ID sendTo chatId

                    return

                }

                targetChat = targetUser.id.toLong()

            } else {

                sudo make L.REPLY_OR_ID sendTo chatId

                return

            }

            val chat = try {

                getChat(targetChat)

            } catch (e: TdException) {

                sudo make L.GET_CHAT_FAILED sendTo chatId

                return

            }

            if (chat.type !is ChatTypePrivate) {

                sudo make L.NO_PRIVATE_WARN sendTo chatId

                return

            }

            sudo.chat = targetChat

            sudo makeHtml L.CHAT_ENTERED.input(chat.title.toInlineMention(chat.id.toInt())) sendTo chatId

        } else if ("exit" == function) {

            if (sudo.chat != -1L) {

                sudo.chat = -1L

                sudo make L.CHAT_EXITED sendTo chatId

            } else {

                sudo make L.NO_CHAT_ENTERED sendTo chatId

            }

        } else if ("chat" == function) {

            val targetChat: Long

            if (message.replyToMessageId != 0L) {

                targetChat = findChat(message)

                if (targetChat == -1L) {

                    sudo make L.INVALID_REPLY sendTo chatId

                    return

                }

            } else if (params.size != 0) {

                val targetUser = findUser(message, param, params)

                if (targetUser == null) {

                    sudo make L.INVALID_CHAT_ID sendTo chatId

                    return

                }

                targetChat = targetUser.id.toLong()

            } else {

                sudo make L.REPLY_OR_ID sendTo chatId

                return

            }

            val session = data.getSession(targetChat)

            val targetUser = getUserOrNull(targetChat.toInt())

            if (targetUser == null) {

                sudo make L.CHAT_NO_RECORD sendTo chatId

                return

            }

            sudo delete message

            sudo makeHtml sessionStat(session) withMarkup sessionActions(session) replyTo message.replyToMessageId sendTo chatId

        }


    }

    suspend fun findUser(message: Message, param: String?, params: Array<String>): User? {
        
        if (NumberUtil.isInteger(param)) {
            
            return getUserOrNull(NumberUtil.parseInt(param))
            
        }
        
        for (entity in (message.content as MessageText).text.entities) {
          
            if (entity.type is TextEntityTypeMentionName) {
                
                val userId = (entity.type as TextEntityTypeMentionName).userId
                
                return getUserOrNull(userId)
                
            }
            
        }
        
        var userName = params[0]
        
        try {
        
            if (userName.startsWith("@")) userName = userName.substring(1)
        
            val chat = searchPublicChat(userName)
        
            if (chat.type is ChatTypePrivate) return getUser((chat.type as ChatTypePrivate).userId)
            
        } catch (ignored: TdException) {
        }
        
        return null
    }

    fun findChat(message: Message): Long {
        
        val actionMessage = message.replyToMessageId
        val actionMessageStr = actionMessage.toString() + ""
        
        if (data.received.containsKey(actionMessageStr)) {
        
            return data.getSessionByElement("received", data.received[actionMessageStr]!!).chatId
        
        } else if (data.sended.containsKey(actionMessageStr)) {
        
            return data.getSessionByElement("sended", data.sended[actionMessageStr]!!).chatId
        
        } else {
        
            var session = data.getSessionByElement("reports", actionMessage)
        
            if (session != null) return session.chatId
  
            session = data.getSessionByElement("extras", actionMessage)
           
            if (session != null) return session.chatId
        
        }
        
        return -1
    
    }

    override suspend fun onStartPayload(userId: Int, chatId: Long, message: Message, payload: String, params: Array<String>) {

        if (bot.owner != userId) return sudo.onLaunch(userId, chatId, message)
        
        val L = userId.langForUserId
        
        sudo delete message

        val sessionId = if (NumberUtil.isLong(params[0])) NumberUtil.parseLong(params[0]) else -1

        val targetUser = getUserOrNull(sessionId.toInt())

        if (targetUser == null) {
            
            sudo make L.CHAT_NO_RECORD sendTo chatId
            
            return
        
        }

        val session = data.getSession(sessionId)

        sudo makeHtml sessionStat(session) withMarkup sessionActions(session) sendTo chatId

    }


    suspend fun sessionStat(session: PmData.Session): String {
        
        val L = bot.owner.langForUserId
        
        var stat: String = L.CHAT_MANAGE_HELP
        
        stat += "\n\n" + L.USER_NAME.toString() + " : " + Fn.mention(getUser(session.chatId.toInt()))

        if (data.blocked.contains(session.chatId)) {
            stat += " " + L.CHAT_IS_BLOCKED
        }

        stat += "\n" + L.USER_ID.toString() + " : " + Fn.code(session.chatId)
        stat += "\n" + L.CHAT_MSG_RECEIVED
        stat += "\n" + L.CHAT_MSG_SENDED

        return StrUtil.format(stat, session.received.size, session.sended.size)

    }

    fun sessionActions(session: PmData.Session) = inlineButton {

        val L = bot.owner.langForUserId

        val sessionId = session.chatId.asByteArray

        dataLine(L.CHAT_MSG_DEL_RECEIVED, DATA_1, 0, sessionId)
        dataLine(L.CHAT_MSG_DEL_SENDED, DATA_1, 1, sessionId)
        dataLine(L.CHAT_MSG_DEL_ALL, DATA_1, 2, sessionId)
        dataLine(if (data.blocked.contains(session.chatId)) L.CHAT_UNBLOCK else L.CHAT_BLOCK, DATA_1, 3, sessionId)
        dataLine(L.FINISH_MANAGE, DATA_1, 4)

    }
    
}